File: kafka-connect-azure-servicebus/src/main/scala/io/lenses/streamreactor/connect/azure/servicebus/source/AzureServiceBusSourceTask.scala
/*
 *
 *  * Copyright 2017 Datamountaineer.
 *  *
 *  * Licensed under the Apache License, Version 2.0 (the "License");
 *  * you may not use this file except in compliance with the License.
 *  * You may obtain a copy of the License at
 *  *
 *  * http://www.apache.org/licenses/LICENSE-2.0
 *  *
 *  * Unless required by applicable law or agreed to in writing, software
 *  * distributed under the License is distributed on an "AS IS" BASIS,
 *  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  * See the License for the specific language governing permissions and
 *  * limitations under the License.
 *
 */

package io.lenses.streamreactor.connect.azure.servicebus.source

import java.util
import java.util.concurrent.ConcurrentHashMap
import java.util.function.BiConsumer

import com.azure.messaging.servicebus.ServiceBusReceivedMessage
import com.datamountaineer.streamreactor.connect.utils.{JarManifest, ProgressCounter}
import com.typesafe.scalalogging.StrictLogging
import io.lenses.streamreactor.connect.azure.servicebus
import io.lenses.streamreactor.connect.azure.servicebus.config.{AzureServiceBusConfig, AzureServiceBusSettings}
import org.apache.kafka.connect.source.{SourceRecord, SourceTask}

import scala.collection.JavaConverters._
import scala.concurrent.duration.{FiniteDuration, MILLISECONDS}

//noinspection VarCouldBeVal
class AzureServiceBusSourceTask extends SourceTask with StrictLogging {

  private val manifest = JarManifest(
    getClass.getProtectionDomain.getCodeSource.getLocation)

  private val progressCounter = new ProgressCounter
  private var enableProgress: Boolean = false
  private val recordsToCommit =
    new ConcurrentHashMap[SourceRecord, MessageAndTimestamp]()
  private var lastEvictedTimestamp: FiniteDuration =
    FiniteDuration(System.currentTimeMillis(), MILLISECONDS)
  private var evictInterval: Int = 0
  //noinspection VarCouldBeVal
  private var evictThreshold: Int = 0
  var reader: AzureServiceBusReader = _
  var nextRead: Long = 0L
  var interval: Long = AzureServiceBusConfig.AZURE_POLL_INTERVAL_DEFAULT

  override def start(props: util.Map[String, String]): Unit = {
    logger.info(
      scala.io.Source
        .fromInputStream(getClass.getResourceAsStream("/servicebus-ascii.txt"))
        .mkString + s" ${version()}")
    logger.info(manifest.printManifest())

    val conf = if (context.configs().isEmpty) props else context.configs()
    AzureServiceBusConfig.config.parse(conf)
    val configBase = AzureServiceBusConfig(conf)
    val settings = AzureServiceBusSettings(configBase)

    interval = settings.pollInterval

    reader = AzureServiceBusReader(
      props.getOrDefault("name", ""),
      settings,
      servicebus.getConverters(settings.converters, conf.asScala.toMap),
      version(),
      manifest.gitCommit(),
      manifest.gitRepo()
    )
    enableProgress =
      configBase.getBoolean(AzureServiceBusConfig.PROGRESS_COUNTER_ENABLED)
  }

  override def poll(): util.List[SourceRecord] = {
    val now = System.currentTimeMillis()

    if (now > nextRead) {
      val records = reader.read().map { r =>
        recordsToCommit.put(
          r.record,
          MessageAndTimestamp(r.message,
            r.source,
            FiniteDuration(now,
              MILLISECONDS)))
        r.record
      }

      if (enableProgress) {
        progressCounter.update(records)
      }
      // increment the time
      nextRead = now + interval
      records.asJava

    } else {
      logger.info(s"Waiting for frequency to pass before polling, [${nextRead - now}] ms remaining")
      List.empty.asJava
    }
  }

  override def stop(): Unit = reader.close()

  override def version(): String = manifest.version()

  // once connect has committed complete the record
  override def commitRecord(record: SourceRecord): Unit = {
    Option(recordsToCommit.remove(record)).foreach {
      case MessageAndTimestamp(msgToken, source, _) =>
        reader.complete(source, msgToken)
    }
    evictUncommittedMessages()
  }

  def getRecordsToCommit: ConcurrentHashMap[SourceRecord, MessageAndTimestamp] =
    recordsToCommit

  private def evictUncommittedMessages(): Unit = {
    val current = FiniteDuration(System.currentTimeMillis(), MILLISECONDS)
    if ((current - lastEvictedTimestamp).toMinutes > evictInterval) {
      recordsToCommit.forEach(
        new BiConsumer[SourceRecord, MessageAndTimestamp] {
          override def accept(t: SourceRecord, u: MessageAndTimestamp): Unit =
            evictIfApplicable(t, u, current)
        })
    }
    lastEvictedTimestamp = current
  }

  private def evictIfApplicable(record: SourceRecord,
                                msg: MessageAndTimestamp,
                                now: FiniteDuration): Unit = {
    if ((now - msg.timestamp).toMinutes > evictThreshold) {
      recordsToCommit.remove(record)
    }
  }

  case class MessageAndTimestamp(message: ServiceBusReceivedMessage,
                                 source: String,
                                 timestamp: FiniteDuration)
}

File: java-connectors/kafka-connect-azure-servicebus/src/main/java/io/lenses/streamreactor/connect/azure/servicebus/source/AzureServiceBusSourceTask.java
/*
 * Copyright 2017-2024 Lenses.io Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.lenses.streamreactor.connect.azure.servicebus.source;

import static io.lenses.streamreactor.common.util.EitherUtils.unpackOrThrow;

import cyclops.control.Option;
import io.lenses.kcql.Kcql;
import io.lenses.streamreactor.common.util.JarManifest;
import io.lenses.streamreactor.connect.azure.servicebus.config.AzureServiceBusConfigConstants;
import io.lenses.streamreactor.connect.azure.servicebus.config.AzureServiceBusSourceConfig;
import io.lenses.streamreactor.connect.azure.servicebus.util.KcqlConfigBusMapper;

import java.time.Duration;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ArrayBlockingQueue;
import java.util.concurrent.atomic.AtomicReference;
import java.util.function.Function;
import lombok.extern.slf4j.Slf4j;
import org.apache.kafka.clients.producer.RecordMetadata;
import org.apache.kafka.common.config.ConfigException;
import org.apache.kafka.connect.errors.ConnectException;
import org.apache.kafka.connect.source.SourceRecord;
import org.apache.kafka.connect.source.SourceTask;

/**
 * Implementation of {@link SourceTask} for Microsoft Azure EventHubs.
 */
@Slf4j
public class AzureServiceBusSourceTask extends SourceTask {

  private final JarManifest jarManifest;
  private TaskToReceiverBridge taskToReceiverBridge;
  private final AtomicReference<Throwable> serviceBusReceiverError = new AtomicReference<>();
  private long sleepOnEmptyPoll = 100;

  public AzureServiceBusSourceTask() {
    this.jarManifest = unpackOrThrow(JarManifest.produceFromClass(getClass()));
  }

  @Override
  public String version() {
    return jarManifest.getVersion();
  }

  @Override
  public void start(Map<String, String> props) {
    final AzureServiceBusSourceConfig config = new AzureServiceBusSourceConfig(props);
    int recordsQueueSize = config.getInt(AzureServiceBusConfigConstants.TASK_RECORDS_QUEUE_SIZE);
    if (recordsQueueSize < 1) {
      throw new ConfigException("Records queue size must be greater than 0");
    }
    sleepOnEmptyPoll = config.getLong(AzureServiceBusConfigConstants.SOURCE_SLEEP_ON_EMPTY_POLL_MS);
    if (sleepOnEmptyPoll < 0) {
      throw new ConfigException("Sleep on empty poll must be greater than or equal to 0");
    }
    String connectionString = config.getPassword(AzureServiceBusConfigConstants.CONNECTION_STRING).value();
    List<Kcql> kcqls =
        KcqlConfigBusMapper.mapKcqlsFromConfig(props.get(AzureServiceBusConfigConstants.KCQL_CONFIG), true)
            .fold(ex -> {
              throw ex;
            }, Function.identity());

    final ArrayBlockingQueue<ServiceBusMessageHolder> recordsQueue =
        new ArrayBlockingQueue<>(recordsQueueSize);

    final int prefetchCount = config.getInt(AzureServiceBusConfigConstants.SOURCE_PREFETCH_COUNT);
    if (prefetchCount < 1) {
      throw new ConfigException("Prefetch count must be greater than 0");
    }
    final int maxCompleteRetries = config.getInt(AzureServiceBusConfigConstants.SOURCE_MAX_COMPLETE_RETRIES);
    if (maxCompleteRetries < 1) {
      throw new ConfigException("Max complete retries must be greater than 0");
    }
    final int completeMinFirstRetryBackoff =
        config.getInt(AzureServiceBusConfigConstants.SOURCE_MIN_BACKOFF_COMPLETE_RETRIES_MS);
    if (completeMinFirstRetryBackoff < 0) {
      throw new ConfigException("Min backoff complete retries must be greater than or equal to 0");
    }
    final Map<String, ServiceBusReceiverFacade> receiversMap =
        ServiceBusReceiverFacadeInitializer.initializeReceiverFacades(recordsQueue, kcqls, connectionString,
            prefetchCount, maxCompleteRetries, Duration.ofMillis(completeMinFirstRetryBackoff),
            serviceBusReceiverError);

    TaskToReceiverBridge serviceBusReceiverBridge =
        new TaskToReceiverBridge(recordsQueue, receiversMap);

    initialize(serviceBusReceiverBridge);
  }

  void initialize(TaskToReceiverBridge taskToReceiverBridge) {
    this.taskToReceiverBridge = taskToReceiverBridge;
  }

  @Override
  public List<SourceRecord> poll() {
    Option.ofNullable(serviceBusReceiverError.get())
        .peek(e -> {
          throw new ConnectException("Error in Service Bus Receiver", e);
        });
    List<SourceRecord> poll =
        taskToReceiverBridge.poll();
    if (poll.isEmpty()) {
      try {
        Thread.sleep(sleepOnEmptyPoll);
      } catch (InterruptedException e) {
        log.trace("Error while sleeping between polling of records", e);
      }
      return null;
    }
    log.debug("Polled {} records", poll.size());
    return poll;
  }

  @Override
  public void commitRecord(SourceRecord committedRecord, RecordMetadata metadata) {
    taskToReceiverBridge.commitRecordInServiceBus(committedRecord, metadata);
  }

  @Override
  public void stop() {
    log.info("Stopping {}", getClass().getSimpleName());
    taskToReceiverBridge.closeReceivers();
  }
}
